### ISSUE TITLE
[no-useless-constructor] should not report on different constructor type

### ISSUE THREAD

**Issue by Zzzen** [typescript-eslint/typescript-eslint#9884]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=MYGwhgzhAEBCkFNoG8BQ1rAPYDsIBcAnAV2Hy0IAoATMfMALmgMIEscBzaAH2h2IC2AIwSEAlCnQZMuCFhAIAdCCwcadMGIDcUgL6p9qUJBgARUawBuCatAQAPfAhzUY8CEjQZseIqXJUtPRMLOwcEl7SEMQADqLq9GJ6BkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0yHJsnxwy%2BDtHRgA2uGwQATIogBdRQF8QWoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class Base {
  constructor(data: string | number) {
    console.log(data);
  }
}

class Derived extends Base {
  constructor(data: string) {
    super(data)
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-useless-constructor": [
      2
    ]
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

no error

### Actual Result

errored on `constructor` of `Derived`

### Additional Info

Actually I want to reopen #3820.
No type info is needed, AST nodes can be directly checked, similar to the approach used in the [no-duplicate-type-constituents check](https://github.com/typescript-eslint/typescript-eslint/blob/a1bcf5bf8aaae5b5c589073095063040c93e45e3/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L19).

**Comment by JoshuaKGoldberg** [comment]
> no type info is needed

```ts
import { Base } from "somewhere-else";

class Derived extends Base {
  constructor(data: string) {
    super(data)
  }
}
```

Type info is needed. 🙂 

> the approach used in the [no-duplicate-type-constituents check](https://github.com/typescript-eslint/typescript-eslint/blob/a1bcf5bf8aaae5b5c589073095063040c93e45e3/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L19).

For reference, that `isSameAstNode` is an optimization for pure AST checks. You can see later in the file that the rule still falls back to using the type checker as a next step: https://github.com/typescript-eslint/typescript-eslint/blob/a1bcf5bf8aaae5b5c589073095063040c93e45e3/packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts#L132

Just confirming before closing this out - am I misreading what you're suggesting?

**Comment by Zzzen** [comment]
I agree that incorporating type information would enhance the accuracy of this rule. However, I suggest that we utilize `isSameAstNode` to minimize the occurrence of false positives.

**Comment by bradzacher** [comment]
@Zzzen that wouldn't minimise anything though because as Josh pointed out - the vast majority of cases are going to be two classes declared in separate files, not declared in the same file.

In order for us to handle cross-file parent classes we need type information.

It's honestly just not worth the implementation effort and maintenance burden to use scope analysis to add handling specifically for the case of "a subclass that is declared in the same file as the parent class which declares the same number of constructor arguments as the parent class, but with different types".

I also don't agree that handling it is a good idea because it creates an asymmetric rule - if you declare the child in the same file as the parent then we handle it, otherwise we don't. It can be confusing to people and lead to issue reports or just mistrust of the lint rules.

**Comment by bradzacher** [comment]
I'm a -1 on this, if it wasn't clear 😅
Cc @typescript-eslint/triage-team

**Comment by JoshuaKGoldberg** [comment]
Yup, same conclusion as before here.
